### ISSUE TITLE
A way to distinguish TextEditingController event types.

### ISSUE THREAD

**Issue by dpedrinha** [flutter/flutter#79982]
## Use case

When using the `TextEditingController.addListener(myListener)` I get events from text changes but also from focus changes.
If I want to change my state, let's say showing an alert at some point when the user is typing, `myListener` is called 2 more times.
Meaning that I need to keep track of the previous text and check every time if the text did actually change or if it was just a focus event.
Same happens when the `TextField` get focused or unfocused. I get calls to `myListener`.

## Proposal

Having two listener types for `TextEditingController`. One for focus events and another one just for text changes.

**Comment by pedromassangocode** [comment]
Related to https://github.com/flutter/flutter/issues/64143.

@justinmc  I'm wondering whether this this is should be considered as an issue rather than a proposal to distinguish event types. From my understanding controller is expected to be notified when the text changes only and not on focus change (in the user perspective). 

I'm considering this a bug for now, feel free to update the labels as required.

Cc @LongCatIsLooong @chunhtai 


The code bellow prints every time the text field gains/loses focus.

<details>
<summary>reproducible code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  late final TextEditingController _controller = TextEditingController()..addListener(() {
    print(_controller.text);
  });

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: Center(
          child: TextField(controller: _controller),
        ),
      ),
    );
  }
}

```

</details>

**Comment by dpedrinha** [comment]
I believe that having two events is better than just removing the focus event because it would give us more control over the textfield. But I'm used to event driven platforms, and flutter isn't that event driven, so the dev team will know what's best.

**Comment by chunhtai** [comment]
I think one of the event is triggered by selection changes? The text editing value includes both selection and text (as well as composing). When a textfield gets focused, the selection changes from TextSelection.collapsed(offset: -1) to TextSelection.collapsed(offset: text.length)

**Comment by justinmc** [comment]
Yeah I think the behavior of calling the listener for both selection and text changes is working as intended, so this is a proposal and not a bug.  [TextEditingValue](https://master-api.flutter.dev/flutter/services/TextEditingValue-class.html) includes both selection and text as @chunhtai said, and since [TextEditingController](https://master-api.flutter.dev/flutter/widgets/TextEditingController-class.html) is just a ValueNotifier<TextEditingValue>, it will call the listener whenever the TextEditingValue changes at all.

The workaround mentioned in the description above: https://dartpad.dartlang.org/eec5ab37140ce52c8eb530b861960907

I wonder if it would be possible to write a subclass of TextEditingController that has `addTextListener` and `addSelectionListener`?

**Comment by LongCatIsLooong** [comment]
IIRC when `EditableText` gets unfocused it sets the selection to `(-1, -1)`. It's probably not needed anymore. Let me see if I can remove it.

**Comment by chunhtai** [comment]
@LongCatIsLooong Be aware the desktop accessibility uses -1 to indicate there is no valid selection.

**Comment by dpedrinha** [comment]
> Yeah I think the behavior of calling the listener for both selection and text changes is working as intended, so this is a proposal and not a bug. [TextEditingValue](https://master-api.flutter.dev/flutter/services/TextEditingValue-class.html) includes both selection and text as @chunhtai said, and since [TextEditingController](https://master-api.flutter.dev/flutter/widgets/TextEditingController-class.html) is just a ValueNotifier, it will call the listener whenever the TextEditingValue changes at all.
> 
> The workaround mentioned in the description above: https://dartpad.dartlang.org/eec5ab37140ce52c8eb530b861960907
> 
> I wonder if it would be possible to write a subclass of TextEditingController that has `addTextListener` and `addSelectionListener`?

Or maybe add a type parameter to the event?

> Just a FYI, there is a onChange parameter for textfield that is specifically for text changes.

Yeah but it's triggered only by human interaction while the TextEditingController event is triggered also when the text is changed programatically. So the onChange doesn't fit all scenarios.

**Comment by LongCatIsLooong** [comment]
> Having two listener types for TextEditingController. One for focus events and another one just for text changes.

It is already possible. You can check in your `TextEditingValue` listener if the text field's `FocusNode` has focus: https://api.flutter.dev/flutter/widgets/FocusNode/hasFocus.html

**Comment by dpedrinha** [comment]
> > Having two listener types for TextEditingController. One for focus events and another one just for text changes.
> 
> It is already possible. You can check in your `TextEditingValue` listener if the text field's `FocusNode` has focus: https://api.flutter.dev/flutter/widgets/FocusNode/hasFocus.html

Well yes, but it's more of a workaround. It doesn't work for all scenarios.

**Comment by LongCatIsLooong** [comment]
I made the focus change problem into a different issue #86752. 

So the other part of the proposal is that `ValueListenable<T>` needs to be able to accept `void Function(T before, T after)` as listeners, so you can filter out uninteresting changes?

**Comment by dpedrinha** [comment]
> I made the focus change problem into a different issue #86752.
> 
> So the other part of the proposal is that `ValueListenable<T>` needs to be able to accept `void Function(T before, T after)` as listeners, so you can filter out uninteresting changes?

It would make it a little easier to check if the text actually changed. But it will still require our management to check `if(after != before)`.

My suggestion is that the event is triggered only `if(after != before)`.

**Comment by LongCatIsLooong** [comment]
Unassigning since the issue I originally intended to address  (#86752) has been addressed.
